diff --git a/stream/stream_radio.c b/stream/stream_radio.c
index d233739..19c536d 100644
--- a/stream/stream_radio.c
+++ b/stream/stream_radio.c
@@ -858,7 +858,7 @@ static int open_s(stream_t *stream,int mode)
     mp_msg(MSGT_RADIO,MSGL_V,"\n");
 
     if(priv->driver)
-        mp_msg(MSGT_RADIO, MSGL_INFO, priv->driver->info);
+        mp_msg(MSGT_RADIO, MSGL_INFO, "%s",priv->driver->info);
     else{
         mp_tmsg(MSGT_RADIO, MSGL_INFO, "[radio] Unknown driver name: %s\n",priv->radio_param->driver);
         close_s(stream);
