--- a/etc/mpv.conf
+++ b/etc/mpv.conf
@@ -1,7 +1,7 @@
 #
 # mpv configuration file
 #
-# Configuration files are read system-wide from /usr/local/etc/mpv.conf
+# Configuration files are read system-wide from /etc/mpv.conf
 # and per-user from ~/.config/mpv/mpv.conf, where per-user settings override
 # system-wide settings, all of which are overridden by the command line.
 #
@@ -27,7 +27,7 @@
 ##################
 
 # Specify default video driver (see --vo=help for a list).
-#vo=opengl-hq
+vo=opengl-hq
 
 # Start in fullscreen mode by default.
 #fs=yes
@@ -49,7 +49,7 @@
 ##################
 
 # Specify default audio driver (see --ao=help for a list).
-#ao=alsa
+ao=pulse,alsa
 
 # Disable softvol usage, and always use the system mixer if available.
 #softvol=no
@@ -102,7 +102,7 @@
 
 # Enable hardware decoding if available. Often, this requires using an certain
 # video output, otherwise no hardware decoding will be used.
-#hwdec=auto
+hwdec=auto
 
 ############
 # Profiles #
